engine/access/rpc/backend/scripts/scripts.go

```diff
@@ -72,67 +116,240 @@ func NewScriptsBackend(
+//   - All errors returned by this API are guaranteed to be benign. The node can continue normal operations after such errors.
+//   - Hence, we MUST check here and crash on all errors *except* for those known to be benign in the present context!
+//
+// Expected error returns during normal operation:
+//   - [access.InvalidRequestError] - the combined size (in bytes) of the script and arguments is greater than the max size or
+//     if the script execution failed due to invalid arguments.
+//   - [access.DataNotFoundError] - if data required to process the request is not available.
+//   - [access.PreconditionFailedError] - if data for block is not available.
+//   - [access.RequestCanceledError] - if the script execution was canceled.
+//   - [access.RequestTimedOutError] - if the script execution timed out.
+//   - [access.ResourceExhausted] - if computation or memory limits were exceeded.
+//   - [access.InternalError] - for internal failures or index conversion errors.
+//   - [access.ServiceUnavailable] - if no nodes are available or a connection to an execution node could not be established.
 func (b *Scripts) ExecuteScriptAtLatestBlock(
 	ctx context.Context,
 	script []byte,
 	arguments [][]byte,
-) ([]byte, error) {
+	userCriteria optimistic_sync.Criteria,
+) ([]byte, *accessmodel.ExecutorMetadata, error) {
 	if !commonrpc.CheckScriptSize(script, arguments, b.maxScriptAndArgumentSize) {
-		return nil, status.Error(codes.InvalidArgument, commonrpc.ErrScriptTooLarge.Error())
+		return nil, nil, access.NewInvalidRequestError(commonrpc.ErrScriptTooLarge)
 	}
 
 	latestHeader, err := b.state.Sealed().Head()
 	if err != nil {
 		// the latest sealed header MUST be available
-		err := irrecoverable.NewExceptionf("failed to lookup sealed header: %w", err)
-		irrecoverable.Throw(ctx, err)
-		return nil, err
+		err = irrecoverable.NewExceptionf("failed to lookup latest sealed header: %w", err)
+		return nil, nil, access.RequireNoError(ctx, err)
 	}
 
-	res, _, err := b.executor.Execute(ctx, executor.NewScriptExecutionRequest(latestHeader.ID(), latestHeader.Height, script, arguments))
-	return res, err
+	latestHeaderID := latestHeader.ID()
+	executionResultInfo, err := b.executionResultProvider.ExecutionResultInfo(
+		latestHeaderID,
+		userCriteria,
+	)
+	if err != nil {
+		err = fmt.Errorf("failed to get execution result info for block: %w", err)
+		switch {
+		case errors.Is(err, storage.ErrNotFound):
+			return nil, nil, access.NewDataNotFoundError("execution data", err)
+		case common.IsInsufficientExecutionReceipts(err):
+			return nil, nil, access.NewDataNotFoundError("execution data", err)
+		default:
+			return nil, nil, access.RequireNoError(ctx, err)
+		}
+	}
+
+	request := executor.NewScriptExecutionRequest(latestHeaderID, latestHeader.Height, script, arguments)
+	res, metadata, err := b.executor.Execute(ctx, request, executionResultInfo)
+	if err != nil {
+		return nil, nil, handleScriptExecutionError(ctx, err)
```

Comment:
this is where the types should be checked and converted to access sentinels. that way I can look at this one function and see that all expected error cases are handled, and exceptions are thrown.